### PR TITLE
KAFKA-10787: Add import ordering checkstyle rule and configure an automatic formatter

### PR DESCRIPTION
**Pull request by gongxuanzhang**
add a import-order rule. It includes the following parts
1. checkstyle import-order 
2. suppress all module 
3. in Intellj IDEA auto format coding configuration

## checkstyle import-order 
In order to uniform coding specifications in Kafka. This is the first step.
Referred to https://github.com/apache/kafka/pull/10428  import order .
We specify the import order as  "kafka, org.apache.kafka, com, net, org, java, javax and static imports."

## suppress all module 
If the detection function is enabled,A lot of code needs to be changed.
In order to pass CI and minimal modification.  suppress all module checkstyle import-order.
We can open the module step by step.

##  in Intellj IDEA auto format coding configuration
If we want to uniform all code. The manual approach is cumbersome and error-prone.
And we need everyone to follow the rules in the future. So we need auto format coding tools.
In this PR I add a IDEA configuration and update `gitignore` file

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,6 +13,7 @@ project/boot/
 project/plugins/project/
 patch-process/*
 .idea
+!/.idea/codeStyles
```
Comment (chia7712) [inline .gitignore]: the whole `idea` folder is not in version control, so do we need this line?

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang could you please include the auto-formater in this PR? That can save our life :)

File: .gitignore (hunk 1 above)
Comment (gongxuanzhang) [reply .gitignore]: > the whole `idea` folder is not in version control, so do we need this line?

It has to be, and it's very important.

This file will allow you to import your idea automatically into the project's code format

Comment (chia7712) [reply .gitignore]: It seems to me using build tool to address that is more acceptable. We all use gradle to build kafka but there are many IDE.

We can add IDE supports in the future but could you please keep this PR simple for now?

**Comment by gongxuanzhang** [comment]
I revised this PR.It add spotless gradle plugin and mini configuration.
It contains two part.
1.  checkstyle import order rule. But suppress all module. you can change `checkstyle/suppressions.xml` open  module check If you want 
2.  auto format import order. Again, I did not open in any module. If you want open the auto format. Just add a module name into `build.gradle #spotlessApplyModules`   like `def spotlessApplyModules = ['core']` .then run `./gradlew spotlessApply` you will get formated code module

File: build.gradle [hunk 2]
```diff
@@ -47,7 +47,7 @@ plugins {
   // Updating the shadow plugin version to 8.1.1 causes issue with signing and publishing the shadowed
   // artifacts - see https://github.com/johnrengelman/shadow/issues/901
   id 'com.github.johnrengelman.shadow' version '8.1.0' apply false
-  id 'com.diffplug.spotless' version '6.14.0' apply false // 6.14.1 and newer require Java 11 at compile time, so we can't upgrade until AK 4.0
+  id 'com.diffplug.spotless' version "${spotlessVersion}" apply false
```
Comment (chia7712) [inline build.gradle]: Do we need this variable? Also, why not using latest version `6.25.0`?

File: build.gradle [hunk 3]
```diff
@@ -1007,7 +1026,7 @@ project(':core') {
     testImplementation libs.junitJupiter
     testImplementation libs.slf4jlog4j
     testImplementation libs.caffeine
-
```
Comment (chia7712) [inline build.gradle]: please revert this change

File: checkstyle/suppressions.xml [hunk 4]
```diff
@@ -361,4 +361,7 @@
     <suppress checks="(ClassDataAbstractionCoupling|ClassFanOutComplexity)"
               files="(ReplicaFetcherThreadBenchmark).java"/>
 
+    <!--import order -->
+    <suppress checks="ImportOrder" files="^*$"/>
```
Comment (chia7712) [inline checkstyle/suppressions.xml]: why we need this change?

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang nice patch. BTW, could you add docs to `README` for spotlessApply?

File: build.gradle [hunk 5]
```diff
@@ -787,6 +800,12 @@ subprojects {
     skipProjects = [ ":jmh-benchmarks", ":trogdor" ]
     skipConfigurations = [ "zinc" ]
   }
+
+  afterEvaluate {
```
Comment (chia7712) [inline build.gradle]: Maybe we can set spotless directly. For example:
```
  if (project.name in spotlessApplyModules) {
    apply plugin: 'com.diffplug.spotless'
    spotless {
      java {
        importOrder('kafka', 'org.apache.kafka', 'com', 'net', 'org', 'java', 'javax', '', '\\#')
        removeUnusedImports()
      }
    }
  }
```

File: build.gradle (hunk 2 above)
Comment (gongxuanzhang) [reply build.gradle]: i write in `gradle.properties` comment, because after 6.13.0 require Java 11 , we should compatibility java 8

File: checkstyle/suppressions.xml (hunk 4 above)
Comment (gongxuanzhang) [reply checkstyle/suppressions.xml]: I think checkstyle should consistent with auto format . If you open the A module auto format, we should open the module check rule.
`ImportOrder` rule can't custom in each module(It's going to take a lot of changes,maybe should add `build.gradle` every module).
So i add this line in order that open rule some module in future,This is what I think is a more convenient way to modify by module

File: build.gradle (hunk 5 above)
Comment (gongxuanzhang) [reply build.gradle]: I change the PR,please review it @chia7712

File: README.md [hunk 6]
```diff
@@ -232,6 +232,12 @@ You can run checkstyle using:
 The checkstyle warnings will be found in `reports/checkstyle/reports/main.html` and `reports/checkstyle/reports/test.html` files in the
 subproject build directories. They are also printed to the console. The build will fail if Checkstyle fails.
 
+#### Spotless ####
+This plugin can review code by rules,it is disabled by default. You can enable it by setting `build.gradle` add a module name into the `spotlessApplyModules` variable like 
+`def spotlessApplyModules = ['core']`,then you can run spotless using:
```
Comment (chia7712) [inline README.md]: `spotlessApplyModules` is a temporary limit, and we don't need to expose it.

File: gradle.properties [hunk 7]
```diff
@@ -28,6 +28,8 @@ scalaVersion=2.13.14
 # Adding swaggerVersion in gradle.properties to have a single version in place for swagger
 # New version of Swagger 2.2.14 requires minimum JDK 11.
 swaggerVersion=2.2.8
+# 6.14.0 and newer are need for java 11
+spotlessVersion=6.13.0
```
Comment (chia7712) [inline gradle.properties]: `spotlessVersion` is used by only one-line, hence we can inline it.

File: checkstyle/suppressions.xml (hunk 4 above)
Comment (chia7712) [reply checkstyle/suppressions.xml]: Making the consistency is extra effort. Also, those changes will complicate checkstyle files. Hence, I prefer to remove them if we do NOT use it actually.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR.

**Comment by gongxuanzhang** [comment]
I update this PR,plz review it

File: build.gradle [hunk 8]
```diff
@@ -47,7 +47,8 @@ plugins {
   // Updating the shadow plugin version to 8.1.1 causes issue with signing and publishing the shadowed
   // artifacts - see https://github.com/johnrengelman/shadow/issues/901
   id 'com.github.johnrengelman.shadow' version '8.1.0' apply false
-  id 'com.diffplug.spotless' version '6.14.0' apply false // 6.14.1 and newer require Java 11 at compile time, so we can't upgrade until AK 4.0
+  //  6.14.0 and newer are need for java 11
```
Comment (chia7712) [inline build.gradle]: how about: `the minimum required JRE of 6.14.0+ is 11`

File: checkstyle/checkstyle.xml [hunk 9]
```diff
@@ -82,6 +82,10 @@
       <property name="file" value="${config_loc}/${importControlFile}"/>
     </module>
 
+    <!--   import order rule not config in the checkstyle.xml.
```
Comment (chia7712) [inline checkstyle/checkstyle.xml]: how about:

```
DON'T define any import order here! Import order check/format is addressed by spotless.
```

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR

File: build.gradle (hunk 8 above)
Comment (gongxuanzhang) [reply build.gradle]: I refer to spotless github:   https://github.com/diffplug/spotless/tree/main/plugin-gradle#requirements

image

File: checkstyle/checkstyle.xml (hunk 9 above)
Comment (gongxuanzhang) [reply checkstyle/checkstyle.xml]: I update that @chia7712

File: README.md [hunk 10]
```diff
@@ -227,11 +227,16 @@ There are two code quality analysis tools that we regularly run, spotbugs and ch
...
 You can run checkstyle using:
 
-    ./gradlew checkstyleMain checkstyleTest
+    ./gradlew checkstyleMain checkstyleTest spotlessCheck
 
 The checkstyle warnings will be found in `reports/checkstyle/reports/main.html` and `reports/checkstyle/reports/test.html` files in the
 subproject build directories. They are also printed to the console. The build will fail if Checkstyle fails.
 
+#### Spotless ####
+This plugin can review code by rules and can also help you check the code , it is disabled by default. Some of our code reviews use `spotless` instead of `checkstyle`:
```
Comment (chia7712) [inline README.md]: `The import order is a part of static check. please call `spotlessApply` to optimize the imports of Java codes before filing pull request.`

Comment (gongxuanzhang) [reply README.md:236]: maybe we will add other check rule by spotless in future ,so i think optimize the imports maybe replaced

**Review by chia7712 — APPROVED** [review]
LGTM!
